### PR TITLE
Add Gmail api

### LINKED ISSUE #258: when connecting to google account, use Gmail API instead of smtp/imap
Some enterprise customers do not support SMTP/IMAP and have it disabled.

We need to use the Gmail API instead when talking to Gmail.

https://developers.google.com/gmail/api/reference/rest

It's what we do on browser extension: https://github.com/FlowCrypt/flowcrypt-browser/blob/master/extension/js/common/api/email-provider/gmail/gmail.ts

When connecting to other email providers, should still use IMAP/SMTP.

### PR DESCRIPTION
**Pull request by Kharchevskyi**
close #258 

Use gmail api to:
- [x] Send message
- [x] Get folders
- [x] Get messages
- [X] Message pagination
- [x] Get single message
- [x] Single message operations (asyncMarkAsReadIfNotAlreadyMarked)
- [x] Single message operations (permanentlyDelete)
- [x] Single message operations (moveToTrash)
- [x] Single message operations (reply)
- [x] Single message operations (archive)
- [x] Single message operations (open 5MB)
- [x] Search messages
- [x] Get isMessageRead status

### CODE REVIEW

**Comment by tomholub** [comment]
I've had trouble building so far - will try to clean build folder & rebuild.

Screenshot 2021-01-26 at 8 22 31 PM

**Comment by tomholub** [comment]
@Kharchevskyi shows the same errors after clean and rebuild. Do I need to update xcode, or something else?

**Comment by tomholub** [comment]
I'll give the update a shot, of it's anything else, let me know

**Comment by Kharchevskyi** [comment]
What version of Xcode do you use? Can you please update to Xcode 12.3, and try to run `bundle exec fastlane do_everything` to update your pods

**Comment by tomholub** [comment]
10.2 - got it - I'll update

**Comment by tomholub** [comment]
Correction, I had 11.6 - will update and re-test.

**Comment by tomholub** [comment]
Updated os and xcode, will try again today

**Comment by tomholub** [comment]
I can build it now. I had to update OS, XCode as well as the iOS device.

**Review by tomholub — APPROVED** [review]
When using the app, it shows in logs:

```
IMAP: creating a new session
SMTP: creating a new session
IMAP AppStartup -> nil 265ms
```

I don't know if that means it's still trying to use imap on Gmail, or if just the logs should be updated.

-----------------

Separately, I don't see any logic to retry API calls when the access token just expired. Is there any such logic? Similar to what is in `Imap+retry`. 

--------------

Overall this looks good -I'll merge this, then we can settle these comments in another PR. I'll create a separate issue.

File: FlowCrypt/Functionality/Mail Provider/Mail Sessions Providers/ConnectionType.swift [hunk 1]
```diff
@@ -9,7 +9,10 @@
 import Foundation
 
 enum AuthType {
-    case oAuth(String), password(String)
+    /// gmail authentication with token
+    case gmail(String)
```
[the remark this answers is not in the document]
Comment (Kharchevskyi) [reply FlowCrypt/Functionality/Mail Provider/Mail Sessions Providers/ConnectionType.swift:13]: Yes, you are right, it was made for purpose to distinguish between providers which we are using, so for example to support Outlook(we are not currently supporting it) we will add a new type `oAuthOutlook(String)`. 
Going to rename it to `oAuthGmail`, cause it's a bit confusing,  indeed

**Pull request merged**